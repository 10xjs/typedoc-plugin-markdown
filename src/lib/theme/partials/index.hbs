{{#if groups}}

{{heading 3}} Index

{{#each groups}}

{{heading 4}} {{title}}

{{#each children}}
* [{{{name}}}]({{relativeUrl url}})
{{/each}}

{{/each}}

{{/if}}