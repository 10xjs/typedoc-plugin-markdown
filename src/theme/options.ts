export class Options {
  public static markdownFlavour: string;
  public static markdownSourcefilePrefix: string;
  public static includes: string;
  public static media: string;
}
