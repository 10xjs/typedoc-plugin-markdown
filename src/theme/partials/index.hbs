## Index

{{#if groups}}
{{#each groups}}
{{#unless isConstructors}}
{{#ifGroupContainesVisibleItems this }}
### {{title}}

{{#each children}}
{{#ifDisplayIndexItem this }}
{{#if url }}
* [{{name}}]({{relativeURL url}})
{{/if}}
{{/ifDisplayIndexItem}}
{{/each}}
{{newLine}}
{{/ifGroupContainesVisibleItems}}
{{/unless}}
{{/each}}
{{/if}}