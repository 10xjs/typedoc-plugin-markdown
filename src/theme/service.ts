/**
 * Global props and util methods
 */
export class ThemeService {

  public static projectName: string;
  public static options: any;
  public static resources: any;

  /**
   * Returns application options
   */
  public static getOptions() {
    return this.options;
  }

  /**
   * Return application resources
   */
  public static getResources() {
    return this.resources;
  }

  /**
   * Returns a compiled partial
   * @param partialName
   * @param data
   */
  public static compilePartial(partialName: string, data: {}) {
    const template = ThemeService.getResources().partials.getResource(partialName).getTemplate();
    return template(data);
  }

  /**
   * Returns compiled template
   * @param templateName
   * @param data
   */
  public static compileTemplate(templateName: string, data: {}) {
    const template = ThemeService.getResources().templates.getResource(templateName).getTemplate();
    return template(data);
  }

  /**
   * Takes an anchor ref and returns id friendly string
   * @param ref
   */
  public static getAnchorRef(ref: string) {
    return ref.replace(/_|\/|\.| /g, '-').replace(/"/g, '').replace(/ /g, '-').toLowerCase();
  }

}
