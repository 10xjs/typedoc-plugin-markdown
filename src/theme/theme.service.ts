import { MarkdownEngine } from './enums/markdown-engine.enum';

/**
 * Global props and util methods
 */
export class ThemeService {

  public static projectName: string;
  public static options: any;
  public static resources: any;

  /**
   * Returns application options
   */
  public static getOptions() {
    return this.options;
  }

  /**
   * Return application resources
   */
  public static getResources() {
    return this.resources;
  }

  /**
   * Returns the specified markdown engine
   */
  public static getMarkdownEngine() {
    const specifiedEngine = this.options.mdEngine || this.options.mdFlavour || MarkdownEngine.GITHUB;
    return specifiedEngine;
  }

  /**
   * Returns a compiled partial
   * @param partialName
   * @param data
   */
  public static compilePartial(partialName: string, data: {}) {
    const template = ThemeService.getResources().partials.getResource(partialName).getTemplate();
    return template(data);
  }

  /**
   * Returns compiled template
   * @param templateName
   * @param data
   */
  public static compileTemplate(templateName: string, data: {}) {
    const template = ThemeService.getResources().templates.getResource(templateName).getTemplate();
    return template(data);
  }

  /**
   * Takes an anchor ref and returns id friendly string
   * @param ref
   */
  public static getAnchorRef(ref: string) {
    return ref.replace(/_|\/|\.| /g, '-').replace(/"/g, '').replace(/ /g, '-').toLowerCase();
  }

}
