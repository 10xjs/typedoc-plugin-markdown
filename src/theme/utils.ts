import * as fs from 'fs';
import * as Handlebars from 'handlebars';
import * as path from 'path';
import { DeclarationReflection, Reflection } from 'typedoc/dist/lib/models/reflections/index';

export function compileTemplate(template: string, data: {}) {
  const source = templateToString(template);
  const md: HandlebarsTemplateDelegate = Handlebars.compile(source);
  return new Handlebars.SafeString(md(data));
}

export function safeString(data: string) {
  return new Handlebars.SafeString(data);
}

export function getAnchorRef(ref: string) {
    return ref.replace(/_|\/|\.| /g, '-').replace(/"/g, '').replace(/ /g, '-').toLowerCase();
  }

function templateToString(template: string) {
  return fs.readFileSync(path.join(__dirname, `/${template}`)).toString();
}
