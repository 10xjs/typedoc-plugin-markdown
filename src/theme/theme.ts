/**
 * Theme definition
 */

/**
 * Typedoc imports
 */
import { DeclarationReflection, ProjectReflection, Reflection } from 'typedoc/dist/lib/models/reflections/index';
import { UrlMapping } from 'typedoc/dist/lib/output/models/UrlMapping';
import { Renderer } from 'typedoc/dist/lib/output/renderer';
import { DefaultTheme } from 'typedoc/dist/lib/output/themes/DefaultTheme';
import { Options } from './options';

interface IOptions {
  markdownFlavour: string;
  markdownSourcefilePrefix: string;
}

export class MarkdownTheme extends DefaultTheme {

  public static buildUrls(reflection: DeclarationReflection, urls: UrlMapping[]): UrlMapping[] {
    const mapping = DefaultTheme.getMapping(reflection);
    if (mapping) {
      const url = [mapping.directory, MarkdownTheme.getUrl(reflection) + '.md'].join('/');
      urls.push(new UrlMapping(url, reflection, mapping.template));
      reflection.url = url;
      reflection.hasOwnDocument = true;
      for (const key in reflection.children) {
        if (reflection.children.hasOwnProperty(key)) {
          const child = reflection.children[key];
          if (mapping.isLeaf) {
            DefaultTheme.applyAnchorUrl(child, reflection);
          } else {
            MarkdownTheme.buildUrls(child, urls);
          }
        }
      }
    } else {
      DefaultTheme.applyAnchorUrl(reflection, reflection.parent);
    }
    return urls;
  }

  public static getUrl(reflection: Reflection, relative?: Reflection, separator: string = '.'): string {

    let url = reflection.getAlias();

    if (reflection.parent && reflection.parent !== relative && !(reflection.parent instanceof ProjectReflection)) {
      url = MarkdownTheme.getUrl(reflection.parent, relative, separator) + separator + url;
    }

    return url;
  }

  private options: IOptions;

  constructor(renderer: Renderer, basePath: string, options: IOptions) {
    super(renderer, basePath);

    this.options = options;

    Options.markdownFlavour = options.markdownFlavour || 'github';
    Options.markdownSourcefilePrefix = options.markdownSourcefilePrefix;

    // remove uneccessary plugins
    renderer.removeComponent('assets');
    renderer.removeComponent('javascript-index');
    renderer.removeComponent('navigation');
    renderer.removeComponent('toc');
    renderer.removeComponent('pretty-print');

  }

  public isOutputDirectory(path: string): boolean {
    return true;
  }

  public getUrls(project: ProjectReflection): UrlMapping[] {
    const urls: UrlMapping[] = [];
    const entryPoint = this.getEntryPoint(project);
    const additionalContext = {
      displayReadme: this.application.options.getValue('readme') !== 'none',
      hideBack: true,
      options: this.options,
    };
    const context = Object.assign(entryPoint, additionalContext);
    // if (this.options.markdownSinglePage) {
    //  urls.push(new UrlMapping('index.md', context, 'reflection.hbs'));
    // } else {
    urls.push(new UrlMapping('index.md', context, 'reflection.hbs'));
    if (entryPoint.children) {
        entryPoint.children.forEach((child: DeclarationReflection) => {
          MarkdownTheme.buildUrls(child, urls);
        });
      // }
    }
    return urls;
  }

}
