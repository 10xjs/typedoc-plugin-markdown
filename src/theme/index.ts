import * as fs from 'fs';
import * as path from 'path';

import { DeclarationReflection, ProjectReflection, Reflection, ReflectionKind } from 'typedoc/dist/lib/models/reflections/index';
import { UrlMapping } from 'typedoc/dist/lib/output/models/UrlMapping';
import { Renderer } from 'typedoc/dist/lib/output/renderer';
import { DefaultTheme } from 'typedoc/dist/lib/output/themes/DefaultTheme';
import { ThemeService } from './service';

export class MarkdownTheme extends DefaultTheme {

  /**
   * This is mostly a copy of the DefaultTheme method with .html ext switched to .md
   * Builds the url for the the given reflection and all of its children.
   *
   * @param reflection  The reflection the url should be created for.
   * @param urls The array the url should be appended to.
   * @returns The altered urls array.
   */
  public static buildUrls(reflection: DeclarationReflection, urls: UrlMapping[]): UrlMapping[] {

    const mapping = DefaultTheme.getMapping(reflection);

    if (mapping) {
      if (!reflection.url || !DefaultTheme.URL_PREFIX.test(reflection.url)) {
        const url = [mapping.directory, DefaultTheme.getUrl(reflection) + '.md'].join('/');
        urls.push(new UrlMapping(url, reflection, mapping.template));
        reflection.url = url;
        reflection.hasOwnDocument = true;
      }
      for (const key in reflection.children) {
        if (reflection.children.hasOwnProperty(key)) {
          const child = reflection.children[key];
          if (mapping.isLeaf) {
            MarkdownTheme.applyAnchorUrl(child, reflection);
          } else {
            MarkdownTheme.buildUrls(child, urls);
          }
        }
      }
    } else {
      MarkdownTheme.applyAnchorUrl(reflection, reflection.parent);
    }

    return urls;
  }

  /**
   * Similar to DefaultTheme method with added functionality to cater for bitbucket heading and single file anchors
   * Generate an anchor url for the given reflection and all of its children.
   *
   * @param reflection  The reflection an anchor url should be created for.
   * @param container   The nearest reflection having an own document.
   */
  public static applyAnchorUrl(reflection: Reflection, container: Reflection) {

    const options = ThemeService.getOptions();

    if (!reflection.url || !DefaultTheme.URL_PREFIX.test(reflection.url)) {

      let anchor = DefaultTheme.getUrl(reflection, container, '.');

      if (reflection['isStatic']) {
        anchor = 'static-' + anchor;
      }

      let anchorRef = '';

      switch (reflection.kind) {
        case ReflectionKind.ExternalModule:
          anchorRef = `external-module-${ThemeService.getAnchorRef(reflection.name)}-`;
          break;
        case ReflectionKind.Class:
          anchorRef = `class-${ThemeService.getAnchorRef(reflection.name)}`;
          break;
        case ReflectionKind.Interface:
          anchorRef = `interface-${ThemeService.getAnchorRef(reflection.name)}`;
          break;
        case ReflectionKind.Module:
          anchorRef = `module-${ThemeService.getAnchorRef(reflection.name)}`;
        case ReflectionKind.Enum:
          const isModule = reflection.parent.kind === 0 || reflection.parent.kind === ReflectionKind.ExternalModule;
          anchorRef = isModule ?
            `module-${ThemeService.getAnchorRef(reflection.name)}` :
            `enumeration-${ThemeService.getAnchorRef(reflection.name)}`;
          break;
        default:
          if (options.mdFlavour === 'bitbucket') {
            let anchorPrefix = '';
            if (reflection.kind === ReflectionKind.ObjectLiteral) {
              anchorPrefix += 'object-literal-';
            }
            reflection.flags.forEach((flag) => {
              anchorPrefix += `${flag}-`;
            });
            const prefixRef = ThemeService.getAnchorRef(anchorPrefix);
            const reflectionRef = ThemeService.getAnchorRef(reflection.name);
            anchorRef = `markdown-header-${prefixRef}${reflectionRef}`;
          } else {
            anchorRef = anchor;
          }

      }

      reflection.url = (container.url !== undefined ? container.url : '') + '#' + anchorRef;
      reflection.anchor = anchor;
      reflection.hasOwnDocument = false;

    }

    reflection.traverse((child: any) => {
      if (child instanceof DeclarationReflection) {
        MarkdownTheme.applyAnchorUrl(child, container);
      }
    });
  }

  constructor(renderer: Renderer, basePath: string, options: any) {
    super(renderer, basePath);

    // remove uneccessary plugins
    renderer.removeComponent('assets');
    renderer.removeComponent('javascript-index');
    renderer.removeComponent('navigation');
    renderer.removeComponent('toc');
    renderer.removeComponent('pretty-print');

    // assign global theme service props
    ThemeService.options = options;
    ThemeService.resources = this.resources;

  }

  /**
   * Test whether the given path contains a documentation generated by this theme.
   *
   * @param path  The path of the directory that should be tested.
   * @returns     TRUE if the given path seems to be a previous output directory,
   *              otherwise FALSE.
   */
  public isOutputDirectory(outPath: string): boolean {
    const files = fs.readdirSync(outPath);
    return fs.existsSync(path.join(outPath, 'README.md')) || (files.length === 1 && path.extname(files[0]) === '.md');
  }

  /**
   * Map the models of the given project to the desired output files.
   *
   * @param project  The project whose urls should be generated.
   * @returns        A list of [[UrlMapping]] instances defining which models
   *                 should be rendered to which files.
   */
  public getUrls(project: ProjectReflection): UrlMapping[] {

    const urls: UrlMapping[] = [];
    const entryPoint = this.getEntryPoint(project);

    ThemeService.projectName = entryPoint.name;

    // pass in additional context
    const additionalContext = {
      displayReadme: this.application.options.getValue('readme') !== 'none',
      hideBreadcrumbs: true,
      isIndex: true,
    };

    const context = { ...entryPoint, ...additionalContext };

    urls.push(new UrlMapping('README.md', context, 'reflection.hbs'));

    if (entryPoint.children) {
      entryPoint.children.forEach((child: DeclarationReflection) => {
        MarkdownTheme.buildUrls(child, urls);
      });
    }

    return urls;
  }

}
