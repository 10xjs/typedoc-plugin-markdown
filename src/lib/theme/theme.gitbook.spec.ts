import * as fs from 'fs-extra';
import { RendererEvent } from 'typedoc/dist/lib/output/events';
import { bootstrapTheme, getNavigationMock } from '../../test/test.utils';
import { GitbookTheme } from './theme.gitbook';

describe(`gitbook theme`, () => {
  let theme: GitbookTheme;
  beforeAll(() => {
    const app = bootstrapTheme([{ name: 'platform', value: 'gitbook' }, { name: 'mode', value: 'modules' }]);
    theme = app.theme as GitbookTheme;
  });
  describe(`summary`, () => {
    test(`should create correct summary output`, () => {
      theme.navigation = getNavigationMock();
      jest.spyOn(fs, 'writeFileSync').mockImplementation();
      jest.spyOn(fs, 'mkdirSync').mockImplementation();
      expect(theme.getSummaryMarkdown(getNavigationMock())).toMatchSnapshot();
    });

    test(`should write summary to file`, () => {
      jest.spyOn(theme, 'getSummaryMarkdown').mockReturnValue('getSummaryMarkdown');
      theme.writeSummary({ outputDirectory: '/outputPath' } as RendererEvent);
      expect(fs.writeFileSync).toHaveBeenCalled();
    });
  });
});
