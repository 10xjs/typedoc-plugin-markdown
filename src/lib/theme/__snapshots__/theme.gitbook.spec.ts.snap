// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gitbook theme summary should create correct summary output 1`] = `
"* [Globals](globals.md)
* [Section](README.md)
  * [Subsection 1](sub-section-1.md)
  * [Subsection 2](sub-section-2.md)"
`;
